feat(payload): abstract payload builder in trait

greged93 opened the pull request:
Abstracts the current `PayloadBuilderHandle` behind a `PayloadBuilder` trait. First step towards being able to have the `PayloadBuilder` be a GAT on the `NodeComponents` trait.

Resolves #10956.

greged93 commented on crates/payload/builder/src/traits.rs:
```diff
@@ -1,9 +1,58 @@
+    Pin<Box<dyn Future<Output = Result<P, PayloadBuilderError>> + Send + Sync>>;
+
+/// A type that can request, subscribe to and resolve payloads.
+#[async_trait::async_trait]
```
I'm not too sure about this, are you ok with using the `async_trait` macro or would you rather returning a future?

mattsse replied:
async_trait is fine here, then we could use this as dyn trait as well.

mattsse commented on crates/payload/builder/src/traits.rs:
```diff
@@ -1,9 +1,58 @@
+
+/// A type that can request, subscribe to and resolve payloads.
+#[async_trait::async_trait]
+pub trait PayloadBuilder {
```
can we move this to payload-primitives perhaps?

mattsse replied:
this should be Send + Unpin

mattsse commented on crates/payload/builder/src/service.rs:
```diff
@@ -141,64 +120,78 @@ where
 
+impl<T> PayloadBuilderHandle<T>
+where
+    T: PayloadTypes + 'static,
```
let's also make `PayloadTypes` 'static

mattsse reviewed (changes requested):
cool, some nits

greged93 commented:
I noticed there were 2 `PayloadBuilderError`, one in `reth-payload-builder` and one in `reth-payload-primitives`. They looked very similar barring the `MissingPayload` variant, so I removed the one from `reth-payload-builder`, and moved the events to primitives as well. This avoids any circular dep between builder and primitives.

mattsse commented on crates/payload/builder/src/lib.rs:
```diff
@@ -28,7 +28,7 @@
 //! use std::task::{Context, Poll};
 //! use reth_payload_builder::{EthBuiltPayload, KeepPayloadJobAlive, EthPayloadBuilderAttributes, PayloadJob, PayloadJobGenerator};
-//! use reth_payload_builder::error::PayloadBuilderError;
+//! use reth_payload_primitives::PayloadBuilderError;
```
we should re-export the PayloadBuilderError in L125 then this wouldn't break too much and some crates don't require an additional dep

mattsse reviewed (changes requested):
very nice, one pedantic nit about re-exports
